cpp/path_module/algorithm/path.cpp

```diff
@@ -1 +1,242 @@
+      relationshipSets.any_rel.insert(rel_type);
+    }
+  }
+}
+
+void Path::DfsByDirection(mgp::Path &path, std::unordered_set<mgp::Relationship> &relationships_set,
+                          const mgp::RecordFactory &record_factory, int64_t path_size, const int64_t min_hops,
+                          const int64_t max_hops, const LabelSets &labelSets, const LabelBoolsStatus &labelStatus,
+                          const RelationshipSets &relationshipSets, bool &any_outgoing, bool &any_incoming,
+                          bool outgoing) {
+  const mgp::Node &node = path.GetNodeAt(path_size);  // get latest node in path, and expand on it
+
+  mgp::Relationships rels = outgoing ? node.OutRelationships() : node.InRelationships();
+  for (auto rel : rels) {
+    // go through every relationship of the node and expand to the other node of the relationship
+    if (relationships_set.find(rel) !=
+        relationships_set.end()) {  // relationships_set contains all relationships already visited in this path, and
+                                    // the usage of this if loop is to evade cycles
+      continue;
+    }
+    mgp::Path cpy = mgp::Path(path);
+    const std::string rel_type = std::string(rel.Type());
+    bool rel_allowed = outgoing ? RelationshipAllowed(rel_type, relationshipSets, any_outgoing, any_incoming, true)
+                                : RelationshipAllowed(rel_type, relationshipSets, any_outgoing, any_incoming, false);
+    if (!rel_allowed) {  // if relationship not allowed, go to next one
+      continue;
+    }
+    cpy.Expand(rel);  // expand the path with this relationships
+    std::unordered_set<mgp::Relationship> relationships_set_cpy = relationships_set;
+    relationships_set_cpy.insert(rel);  // insert the relationship into visited relationships
+
+    /*this part is for label filtering*/
+    LabelBools labelBools;
+    mgp::Labels labels = outgoing ? rel.To().Labels() : rel.From().Labels();
+    for (auto label : labels) {  // set booleans to their value for the label of the finish node
+      FilterLabel(label, labelSets, labelBools);
+    }
+
+    if (PathSizeOk(path_size, max_hops, min_hops) && !labelBools.blacklisted &&
+        ((labelBools.end_node && labelStatus.end_node_activated) || labelBools.terminated ||
+         (!labelStatus.termination_activated && !labelStatus.end_node_activated &&
+          Whitelisted(labelBools.whitelisted, labelStatus.whitelist_empty)))) {
+      auto record = record_factory.NewRecord();
+      record.Insert(std::string(std::string(kResultExpand).c_str()).c_str(), cpy);
+    }
+    if (path_size + 1 < max_hops && !labelBools.blacklisted && !labelBools.terminated &&
+        (labelBools.end_node || Whitelisted(labelBools.whitelisted, labelStatus.whitelist_empty))) {
+      Path_DFS(cpy, relationships_set_cpy, record_factory, path_size + 1, min_hops, max_hops, labelSets, labelStatus,
```

Comment:
I see that `cpy` is passed by reference - is the algorithm still correct if there are multiple iterations of the `for` loop that call `Path_DFS`?

Reply:
Yes, it is correct, when cpy is passed by reference, if the algorithm chooses to expand in the next iteration, it will make a new copy and pass its reference